### PR TITLE
bpo-46162: make `property` generic

### PR DESCRIPTION
**Pull request by sobolevn**
Original discussion: https://github.com/python/typing/issues/985

https://bugs.python.org/issue46162

Related, `cached_property` is already generic: https://github.com/python/cpython/blob/main/Lib/functools.py#L1002

### CODE REVIEW

**Review by kumaraditya303 — APPROVED** [review]
LGTM but the PEP should be changed first to account for this change.

**Comment by sobolevn** [comment]
@kumaraditya303 this was my exact question. Because, current version of PEP states:

> This is why starting with Python 3.9, the following collections become generic using __class_getitem__() to parameterize contained types:

I have several issues with that:
- Python3.9 won't support (without `__future__` import) generic `property[]` as far as I understand (I might be wrong here)
- `property` is not a "collection" (as many other types there)

**Comment by isidentical** [comment]
> LGTM but the PEP should be changed first to account for this change.

@kumaraditya303 I am not sure how did you end up with this idea. PEP's are historical documents, and they don't get updated as the language grows. There were multiple occurrences of generic support for classes that the PEP initially did not target (the most recent one I recall is `graphlib.TopologicalSorter`).

@sobolevn  You can simply wait for a few core developers to respond, or just post this on the python-dev (since this is more of a change in the builtins).

File: Lib/test/test_property.py [hunk 1]
```diff
@@ -214,6 +214,21 @@ def test_property_set_name_incorrect_args(self):
...
+        self.assertIs(p.__origin__, property)
+        self.assertEqual(p.__args__, (int, str))
+        self.assertEqual(p.__parameters__, ())
+
+        from typing import TypeVar
+        G = TypeVar('G')
+        S = TypeVar('S')
+        p1 = property[G, S]
+        self.assertEqual(p1.__args__, (G, S))
+        self.assertEqual(p1.__parameters__, (G, S))
```
Comment (isidentical) [inline Lib/test/test_property.py]: Please see `Lib/test_genericalias.py` as well for the tests.

**Comment by kumaraditya303** [comment]
> I am not sure how did you end up with this idea. PEP's are historical documents, and they don't get updated as the language grows. There were multiple occurrences of generic support for classes that the PEP initially did not target (the most recent one I recall is `graphlib.TopologicalSorter`).

This changes `property` which is in the builtins which is very widely used and it not in other module so it should be mentioned in the PEP under 3.11 changes for maximum visibility about this change.

**Comment by isidentical** [comment]
> This changes property which is in the builtins which is very widely used and it not in other module so it should be mentioned in the PEP under 3.11 changes for maximum visibility about this change.

We *generally* do not change old PEPs, especially when they are in the *Final* state. A new discussion can (or should [depending on the complexity of this feature]) be brought up on python-dev, though for a case like this I'd personally suggest just waiting the final resolution in python/typing#985 and also the responses of other developers interested in typing before making an action towards python-dev since it might evolve a bit (which I personally do not expect).

File: Misc/NEWS.d/next/Library/2021-12-23-12-53-34.bpo-46162.bE3xrz.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Make :class:`property` generic.
```
Comment (merwok) [inline Misc/NEWS.d/next/Library/2021-12-23-12-53-34.bpo-46162.bE3xrz.rst:1]: _generic_ can be a bit vague (or generic 😉), could you rephrase this news snippet and the PR title to clarify this is about typing generic alias?
